### PR TITLE
[exporter/datasetexporter]: Initial Commit

### PR DESCRIPTION
**Pull request by martin-majlis-s1**
**Description:** Adding exporter DataSetExporter - initial commit

**Link to tracking Issue:** #20660

**Testing:** This is initial commit without any implementation. It contains only the structure of the exporter.

**Documentation:** Initial README.md describing configuration options.

### CODE REVIEW

**Comment by martin-majlis-s1** [comment]
> [CLA Missing ID](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/196414933/20733/#/?version=2) [CLA Not Signed](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/196414933/20733/#/?version=2)
> 
> * ❌ The commit ([015ca51](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/015ca512d63caf55b8f27c003dc041683af5b94e)). This user is missing the User's ID, preventing the EasyCLA check. [Consult GitHub Help](https://help.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-linked-to-the-wrong-user) to resolve.For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).

I am working on signing EasyCLA.

**Comment by martin-majlis-s1** [comment]
@atoulme is the sponsor of the linked issue.

File: exporter/elasticsearchexporter/config.go
Comment (atoulme) [inline exporter/elasticsearchexporter/config.go:101]:
```suggestion
	// APIKey is used to configure ApiKey based Authentication.
```

**Comment by atoulme** [comment]
This is the right scope of change, thank you! We will need to deal with a few administrative changes to the repo:
* [ ] run `make gendependabot`
* [ ] run `make genotelcontribcol`
* [ ] run `make generate-gh-issue-templates`

**Comment by martin-majlis-s1** [comment]
@atoulme : I have reverted those changes in elastic search, since they are not related to dataset. It was automatically changed during refactoring and I have missed it.

**Comment by martin-majlis-s1** [comment]
@atoulme : I have run those commands - and two of them looks problematic to me:

## ✅  Validate Changelog

`make chlog-validate`
```
cd .../opentelemetry-collector-contrib/internal/tools && go build -o .../opentelemetry-collector-contrib/.tools/chloggen -trimpath go.opentelemetry.io/build-tools/chloggen
.../opentelemetry-collector-contrib/.tools/chloggen validate
PASS: all files in .../opentelemetry-collector-contrib/.chloggen/ are valid
```

## ❌  Run Dependabot

`make gendependabot`
```
Recreating .github/dependabot.yml file
Add entry for "/" github-actions
Add entry for "/" docker
Add entry for "/" gomod
Add entry for "/cmd/configschema"
Add entry for "/cmd/mdatagen"
...
Add entry for "/receiver/webhookeventreceiver"
Add entry for "/receiver/windowseventlogreceiver"
Add entry for "/receiver/windowsperfcountersreceiver"
The following modules are not included in the dependabot file because it has a limit of 220 entries:
  - /receiver/zipkinreceiver
  - /receiver/zookeeperreceiver
  - /testbed
  - /testbed/mockdatareceivers/mockawsxrayreceiver
  - /testbed/mockdatasenders/mockdatadogagentexporter
```

2 existing receivers were removed and more of them in the future

## ❌  Gen Contrib

`make genotelcontribcol`
```
cd .../opentelemetry-collector-contrib/internal/tools && go build -o .../opentelemetry-collector-contrib/.tools/builder -trimpath go.opentelemetry.io/collector/cmd/builder
.../opentelemetry-collector-contrib/.tools/builder --skip-compilation --config cmd/otelcontribcol/builder-config.yaml --output-path cmd/otelcontribcol
Flag --output-path has been deprecated, use config distribution::output_path
2023-04-10T18:04:06.219+0200    INFO    internal/command.go:126 OpenTelemetry Collector Builder {"version": "dev", "date": "unknown"}
2023-04-10T18:04:06.225+0200    INFO    internal/command.go:159 Using config file       {"path": "cmd/otelcontribcol/builder-config.yaml"}
2023-04-10T18:04:06.225+0200    INFO    builder/config.go:116   Using go        {"go-executable": "/opt/homebrew/bin/go"}
2023-04-10T18:04:06.233+0200    INFO    builder/main.go:76      Sources created {"path": "cmd/otelcontribcol"}
Error: failed to update go.mod: exit status 1. Output:
go: downloading github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datasetexporter v0.75.0
github.com/open-telemetry/opentelemetry-collector-contrib/cmd/otelcontribcol imports
        github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datasetexporter: reading github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datasetexporter/go.mod at revision exporter/datasetexporter/v0.75.0: unknown revision exporter/datasetexporter/v0.75.0
```

## ✅ Gen GH Issue
`make generate-gh-issue-templates`
```
for FILE in bug_report feature_request other; do \
                YAML_FILE=".github/ISSUE_TEMPLATE/${FILE}.yaml"; \
                TMP_FILE=".github/ISSUE_TEMPLATE/${FILE}.yaml.tmp"; \
                cat "${YAML_FILE}" > "${TMP_FILE}"; \
                FILE="${TMP_FILE}" ./.github/workflows/scripts/add-component-options.sh > "${YAML_FILE}"; \
                rm "${TMP_FILE}"; \
        done
```

**Comment by martin-majlis-s1** [comment]
Follow up PR is #20870

**Comment by atoulme** [comment]
Changes look good to me. You need to run tidy:
```
make for-all CMD="make tidy"
```

File: .github/CODEOWNERS
Comment (martin-majlis-s1) [reply .github/CODEOWNERS:49]: Thanks, I have created the issue - https://github.com/open-telemetry/community/issues/1429.

**Comment by martin-majlis-s1** [comment]
> Changes look good to me. You need to run tidy:
> 
> ```
> make for-all CMD="make tidy"
> ```

Thanks. I was wondering whether really "every" PR shuffles those files.

File: exporter/datasetexporter/README.md
Comment (atoulme) [inline exporter/datasetexporter/README.md:65]: you don't need those links down there anymore, since they're generated in the metadata status table.

**Comment by martin-majlis-s1** [comment]
Although I run `make generate` - and based on this PR - https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/20264 - I should update it, there is no change needed.

**Review by atoulme — APPROVED** [review]

**Comment by martin-majlis-s1** [comment]
https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4720002829/jobs/8386253303?pr=20733

I have executed: `make -j2 golint GROUP=other`

```
exporters_test.go:24: File is not `gci`-ed with --skip-generated -s standard,default,prefix(github.com/open-telemetry/opentelemetry-collector-contrib) (gci)
        "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datasetexporter"
exporters_test.go:48: File is not `gci`-ed with --skip-generated -s standard,default,prefix(github.com/open-telemetry/opentelemetry-collector-contrib) (gci)
        "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter"
exporters_test.go:28: File is not `goimports`-ed with -local github.com/open-telemetry/opentelemetry-collector-contrib (goimports)
```

So I have executed: `gci write --skip-generated -s standard -s default -s "prefix(github.com/open-telemetry/opentelemetry-collector-contrib)" cmd/otelcontribcol/exporters_test.go`

And `make -j2 golint GROUP=other` passes.

**Comment by martin-majlis-s1** [comment]
https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4720002829/jobs/8386252076?pr=20733 - this workflow fails because `make genotelcontribcol` - believes that generated files are out of date.

When I merge latest master and rerun that command, no changes are made. So I was not able to fix it.

File: .github/dependabot.yml
Comment (evan-bradley) [inline .github/dependabot.yml:1100]: Was this change intentional?

File: exporter/datasetexporter/README.md
Comment (evan-bradley) [inline exporter/datasetexporter/README.md:20]: It could be worth considering putting the settings into a table, [here's an example](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver#configuration). What do you think?

File: exporter/datasetexporter/config_test.go
Comment (evan-bradley) [inline exporter/datasetexporter/config_test.go:23]: The guidance for testing libraries suggests using [stretchr/testify](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#testing-library-recommendations) for anything needed beyond the Go standard library for suites/assertions/etc. Ideally we'd be consistent with this across the Collector, but the advice in the Collector core isn't a hard requirement. Is there an advantage to using this library over testify?

File: exporter/datasetexporter/factory.go
Comment (evan-bradley) [inline exporter/datasetexporter/factory.go:36]: The issue and readme only mention logs. Is the intention to support metrics and traces as well?

**Review by evan-bradley — COMMENTED** [review]
Thanks for adding this, looking pretty good so far. Left a couple of questions/suggestions.

File: exporter/datasetexporter/README.md
Comment (martin-majlis-s1) [reply exporter/datasetexporter/README.md:20]: I have randomly clicked on 5 exporters - https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter - and I used the same style that everybody else.

If table is new way, I can update it.

File: exporter/datasetexporter/config_test.go
Comment (martin-majlis-s1) [reply exporter/datasetexporter/config_test.go:23]: It was used by github.com/jarcoal/httpmock that I am using in the tests for the library - https://github.com/scalyr/dataset-go - that will be used by this exporter. I can swap it.

File: exporter/datasetexporter/factory.go
Comment (martin-majlis-s1) [reply exporter/datasetexporter/factory.go:36]: For now I have follow up PR, where I am supporting logs and traces - https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/20870. 

I can adjust it in this PR or I can do it in the real PR.

File: exporter/datasetexporter/README.md
Comment (evan-bradley) [reply exporter/datasetexporter/README.md:20]: I don't think there's an official or recommended way to do it. I like the structure tables offer, which is why I suggested it, but you're free to choose whatever format you prefer; they both offer the same information.

File: .github/dependabot.yml
Comment (evan-bradley) [reply .github/dependabot.yml:1100]: Makes sense. Thanks!

File: exporter/datasetexporter/factory.go
Comment (evan-bradley) [reply exporter/datasetexporter/factory.go:36]: I think it makes sense to add the implementation in that PR, I just checking since the metadata file and linked issue only mentioned logs. So long as this is in sync with what is in the metadata file [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/20733/files#diff-15048bd70a8a5544947470c6c938c9e0d95e3a2aca2e18a74c24b7eede632770R6) I think we're good.

Comment (martin-majlis-s1) [reply exporter/datasetexporter/factory.go]: I have removed metrics in this PR and updated the metadata, so it should be in sync.

**Comment by atoulme** [comment]
Please resolve the lint issues:
```
Error: datasetexporter.go:85:27: func `(*datasetExporter).consumeMetrics` is unused (unused)
func (e *datasetExporter) consumeMetrics(ctx context.Context, ld pmetric.Metrics) error {
                          ^
make[2]: *** [../../Makefile.Common:154: lint] Error 1
```

**Comment by martin-majlis-s1** [comment]
@atoulme: I have fixed it. :/ 

I am always executing several different commands before I commit these changes and I miss lint. It would be great if these simple workflows could be executed before the approval step.

**Review by atoulme — APPROVED** [review]

**Comment by martin-majlis-s1** [comment]
🥳 🎉 Finally ✅ 

So the current status is:
* waiting for the membership approval - https://github.com/open-telemetry/community/issues/1429
* waiting on @Aneurysm9 to take a look
* waiting on @evan-bradley to take a look again

Is this correct?  

I will keep eye on this PR and I will try to periodically resolve potential conflicts, so that we can merge it once it's ready.

**Comment by atoulme** [comment]
We can merge this without waiting for your membership approval imo. This PR is open to review from anyone, with a maintainer ultimately responsible to perform the merge.

**Comment by martin-majlis-s1** [comment]
I have replaced testing library with `github.com/stretchr/testify/` library.

I have also executed following commands:
```
make gendependabot
make chlog-validate
make genotelcontribcol
make generate-gh-issue-templates
make for-all CMD="make tidy"
make generate
(cd exporter/datasetexporter; make)
(cd exporter/datasetexporter; make test)
make -j2 golint GROUP=other
```

cc: @atoulme , @evan-bradley

File: exporter/datasetexporter/config_test.go
Comment (evan-bradley) [inline exporter/datasetexporter/config_test.go:1]: FYI in testify, the assertion functions usually have function signatures with `(expected, equal)`. You'll want to swap the parameters around in your assertions or the error messages will be confusing if the tests fail.

File: exporter/datasetexporter/config_test.go
Comment (evan-bradley) [inline exporter/datasetexporter/config_test.go:148]:
```suggestion
```
I think you can remove this, too.

File: exporter/datasetexporter/config_test.go
Comment (evan-bradley) [inline exporter/datasetexporter/config_test.go:55]:
```suggestion
```
I'd say you can just remove this check. You could maybe use `EqualError` or `EqualValues` for the assertion above, but that's mostly stylistic. I think what you're already checking works.

File: exporter/datasetexporter/datasetexporter.go
Comment (evan-bradley) [inline exporter/datasetexporter/datasetexporter.go:63]:
```suggestion
```

**Review by evan-bradley — APPROVED** [review]
Thanks for your contribution and your patience throughout the review process for this PR. Most of my suggestions are nonfunctional improvements to the tests, overall this looks good to me.

File: exporter/datasetexporter/config_test.go
Comment (martin-majlis-s1) [reply exporter/datasetexporter/config_test.go:1]: Now I recall, that when I was switching from testify to deepest, I was swapping order. Now I have just replaced those functions. :/

I will fix this. Good catch.

File: exporter/datasetexporter/config_test.go
Comment (martin-majlis-s1) [reply exporter/datasetexporter/config_test.go:32]: I have needed the suite because of the http mocking library, that I was using. I will see in the next PR, if it's needed or not.

**Comment by martin-majlis-s1** [comment]
I have swapped the order of the parameters and accepted suggestions by @evan-bradley .

Then I have executed following commands:
```
make gendependabot
make chlog-validate
make genotelcontribcol
make generate-gh-issue-templates
make for-all CMD="make tidy"
make generate
(cd exporter/datasetexporter; make)
(cd exporter/datasetexporter; make test)
make -j2 golint GROUP=other
```

**Review by evan-bradley — APPROVED** [review]

**Comment by martin-majlis-s1** [comment]
There are two failed workflows - https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4801250488/jobs/8544934628?pr=20733

* https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4801250488/jobs/8544130549?pr=20733 - there was new release of open telemetry, so there was mismatch in version check:
  ```
  Incorrect version "v0.75.0" of "go.opentelemetry.io/collector " is included in "./exporter/datasetexporter/go.mod". It should be version "v0.75.1-0.20230424163446-8e33ded10872"
  ```
  * I have executed `.github/workflows/scripts/check-collector-module-version.sh` to fix all the problems that are detected after the new release.
* https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4801250488/jobs/8544934628?pr=20733 - it's not possible to build deb package, since it fails on:
  ```
  The command '/bin/sh -c apt-get update &&    apt-get install -yq ca-certificates procps systemd' returned a non-zero code: 100
  ```
  * I am not trying to fix it and I hope that next time it succeeds

**Comment by evan-bradley** [comment]
Thanks for your continued patience here, we've been having some CI issues. I've started another run, we'll see if everything is fixed.

**Comment by martin-majlis-s1** [comment]
When I try to run those commands locally: `make docker-otelcontribcol` or `make gotest GROUP=other` - they succeed
```

**Comment by atoulme** [comment]
Try to make sure you rebase off main: `git pull --rebase origin main`. Unfortunately we had a rough week and we're finishing up the release. I will check the main branch CI.

**Comment by martin-majlis-s1** [comment]
I have merged latest master and I have executed following commands:

```sh
make gendependabot && \
make chlog-validate && \
make genotelcontribcol && \
make generate-gh-issue-templates && \
make for-all CMD="make tidy" && \
make generate && \
(cd exporter/datasetexporter; make) && \
(cd exporter/datasetexporter; make test) && \
make -j2 golint GROUP=other && \
.github/workflows/scripts/check-collector-module-version.sh
```

**Comment by martin-majlis-s1** [comment]
It has failed on the step: `make -j2 goporto`, because:
```
--- a/internal/components/components.go
+++ b/internal/components/components.go
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package components // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/components"
+package components // import "opentelemetry-collector-contrib/internal/components"
 
 import (
 	"go.opentelemetry.io/collector/exporter"
```

https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4859088452/jobs/8662599627?pr=20733


I have merged master and I am running following commands:

```bash
make gendependabot && \
make chlog-validate && \
make genotelcontribcol && \
make generate-gh-issue-templates && \
make generate && \
(cd exporter/datasetexporter; make) && \
(cd exporter/datasetexporter; make test) && \
make -j2 golint GROUP=other && \
.github/workflows/scripts/check-collector-module-version.sh && \
make -j2 goporto && \
make -j2 crosslink && \
make -j2 gotidy && \
make -j2 generate && \
 make multimod-verify && \
make for-all CMD="make tidy"
```

Screenshot 2023-05-02 at 17 12 37

File: go.mod
Comment (atoulme) [inline go.mod:650]: This probably derailed CI.

Comment (martin-majlis-s1) [reply go.mod:651]: This is mystery for me, why are here any changes at all.

Comment (martin-majlis-s1) [reply go.mod]: I have replaced it with version from main. :/

**Comment by martin-majlis-s1** [comment]
I have again resolved all the merge conflicts and rerun following commands:
```bash
make gendependabot && \
make chlog-validate && \
make genotelcontribcol && \
make generate-gh-issue-templates && \
make generate && \
(cd exporter/datasetexporter; make) && \
(cd exporter/datasetexporter; make test) && \
make -j2 golint GROUP=other && \
.github/workflows/scripts/check-collector-module-version.sh && \
make -j2 goporto && \
make -j2 crosslink && \
make -j2 gotidy && \
make -j2 generate && \
 make multimod-verify && \
make for-all CMD="make tidy"
```

**Comment by martin-majlis-s1** [comment]
It takes me 40 minutes to resolve all those conflicts:
Screenshot 2023-05-03 at 20 41 26

**Comment by martin-majlis-s1** [comment]
There is again merge conflict 🙈 
Screenshot 2023-05-04 at 9 36 19

**Comment by atoulme** [comment]
You're very close :)

**Comment by martin-majlis-s1** [comment]
I have merged main and executed: `make gendependabot` to fix the merge conflict.

**Comment by enykeev** [comment]
Can we enable auto-merge so that next time this PR is approved and the checks pass, it immediately get merged in?

**Review by atoulme — APPROVED** [review]

File: exporter/datasetexporter/config.go
Comment (TylerHelmuth) [inline exporter/datasetexporter/config.go:26]: Does this need exported?

File: exporter/datasetexporter/README.md
Comment (martin-majlis-s1) [reply exporter/datasetexporter/README.md:8]: Thank you for taking the time to review my pull request. I appreciate your feedback. But could you, please, clarify what should I do?

Guideline for [contributing](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-new-components) - says: "The component should use [In Development Stability](https://github.com/open-telemetry/opentelemetry-collector#development) in its README".

File: exporter/datasetexporter/config.go
Comment (martin-majlis-s1) [reply exporter/datasetexporter/config.go:26]: Good eyes! I will fix it.

File: exporter/datasetexporter/README.md
Comment (martin-majlis-s1) [reply exporter/datasetexporter/README.md:8]: Thanks for clarifying!

Comment (martin-majlis-s1) [reply exporter/datasetexporter/README.md]: I have updated `matadata.yaml` to the new syntax, executed `make generate` and adjusted the code.

**Comment by martin-majlis-s1** [comment]
I have again executed these commands:

```
make gendependabot && \
make chlog-validate && \
make genotelcontribcol && \
make generate-gh-issue-templates && \
make generate && \
(cd exporter/datasetexporter; make) && \
(cd exporter/datasetexporter; make test) && \
make -j2 golint GROUP=other && \
.github/workflows/scripts/check-collector-module-version.sh && \
make -j2 goporto && \
make -j2 crosslink && \
make -j2 gotidy && \
make -j2 generate && \
 make multimod-verify && \
make for-all CMD="make tidy"
```

And all of them passes and no changes are made.

**Comment by martin-majlis-s1** [comment]
Thanks everyone for your time and feedback! I appreciate your patience and expertise, and I'm grateful for the opportunity to learn from you all. Looking forward to contributing more in the future.

**Comment by martin-majlis-s1** [comment]
There is follow up PR - https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21513

**Pull request merged**
